hal-core/src/mem/page/alloc.rs

```diff
@@ -0,0 +1,571 @@
+                "size is less than the minimum page size; rounding up"
+            );
+            size = self.min_size;
+        }
+
+        // Is the size a power of two?
+        if !size.is_power_of_two() {
+            let rounded = size.next_power_of_two();
+            tracing::warn!(
+                size,
+                rounded,
+                allocation.page_size = page_size.as_usize(),
+                allocation.len = len,
+                "size is not a power of two; rounding up"
+            );
+            size = rounded;
+        }
+
+        // Is there enough room to meet this allocation request?
+        let available = self.heap_size.load(Acquire);
+        if size > available {
+            tracing::error!(
+                size,
+                available,
+                allocation.page_size = page_size.as_usize(),
+                allocation.len = len,
+                "out of memory!"
+            );
+            return Err(AllocErr::oom());
+        }
+
+        Ok(size)
+    }
+
+    /// Returns the order of the block that would be allocated for a range of
+    /// `len` pages of size `size`.
+    fn order_for<S: Size>(&self, page_size: S, len: usize) -> Result<usize> {
+        self.size_for(page_size, len)
+            .map(|size| self.order_for_size(size))
+    }
+
+    /// Returns the order of a block of `size` bytes.
+    fn order_for_size(&self, size: usize) -> usize {
+        size.log2() - self.min_size_log2
+    }
+}
+
+impl<L> BuddyAlloc<L>
+where
+    L: AsRef<[spin::Mutex<List<Free>>]>,
+{
+    /// Adds a memory region to the heap from which pages may be allocated.
+    #[tracing::instrument(skip(self), level = "trace")]
+    pub unsafe fn add_region(&self, mut region: Region) -> core::result::Result<(), ()> {
+        // Is the region in use?
+        if region.kind() != RegionKind::FREE {
+            tracing::warn!(?region, "cannot add to page allocator, region is not free");
+            return Err(());
+        }
+
+        let size = region.size();
+        let base = region.base_addr();
+        tracing::info!(region.size = size, region.base_addr = ?base, "adding region");
+
+        // Is the region aligned on the heap's minimum page size? If not, we
+        // need to align it.
+        if !base.is_aligned(self.min_size) {
+            let new_base = base.align_up(self.min_size);
+            tracing::trace!(region.new_base = ?new_base, "base address not aligned!");
+            region = Region::new(new_base, region.size(), RegionKind::FREE);
+        }
+
+        // Is the size of the region a power of two? The buddy block algorithm
+        // requires each free block to be a power of two.
+        if !size.is_power_of_two() {
+            // If the region is not a power of two, split it down to the nearest
+            // power of two.
+            let prev_power_of_two = prev_power_of_two(size);
+            tracing::debug!(prev_power_of_two, "not a power of two!");
```

Comment:
Won't `prev_power_of_two` be a power of two? It might be more useful to log out `size` here.

Reply:
we already logged the size of the region a few lines ago:
https://github.com/hawkw/mycelium/blob/aaaf737e1fd92d11a5d0fa01df0b2d4e3c70555d/hal-core/src/mem/page/alloc.rs#L202